pennylane/beta/tapes/rev.py

```diff
@@ -0,0 +1,204 @@
+        # before each Jacobian call, so that the statevector is calculated only once.
+        self._state = None
+        return super().jacobian(device, params, **options)
+
+    def analytic_pd(self, idx, device, params, **options):
+        t_idx = list(self.trainable_params)[idx]
+        op = self._par_info[t_idx]["op"]
+        p_idx = self._par_info[t_idx]["p_idx"]
+
+        # The reversible tape only support differentiating
+        # expectation values of observables for now.
+        for m in self.measurements:
+            if (
+                m.return_type is qml.operation.Variance
+                or m.return_type is qml.operation.Probability
+            ):
+                raise ValueError(
+                    f"{m.return_type} is not supported with the reversible gradient method"
+                )
+
+        # The reversible tape only supports the RX, RY, RZ, and Rot operations for now:
+        #
+        # * CRX, CRY, CRZ ops have a non-unitary matrix as generator.
+        #
+        # * PauliRot, MultiRZ, U2, and U3 do not have generators specified.
+        #
+        # TODO: the controlled rotations can be supported by multiplying ``state``
+        # directly by these generators within this function
+        # (or by allowing non-unitary matrix multiplies in the simulator backends)
+
+        if op.name not in ["RX", "RY", "RZ", "Rot"]:
+            raise ValueError(
+                "The {} gate is not currently supported with the "
+                "reversible gradient method.".format(op.name)
+            )
+
+        if self._state is None:
+            self.execute_device(params, device)
+            self._state = device._pre_rotated_state
+
+        self.set_parameters(params)
+
+        # create a new circuit which rewinds the pre-measurement state to just after `op`,
+        # applies the generator of `op`, and then plays forward back to
+        # pre-measurement step
+        wires = op.wires
+        op_idx = self.operations.index(op)
+
+        # TODO: likely better to use circuitgraph to determine minimally necessary ops
+        between_ops = self.operations[op_idx + 1 :]
+
+        if op.name == "Rot":
+            decomp = op.decomposition(*op.parameters, wires=wires)
+            generator, multiplier = decomp[p_idx].generator
+            between_ops = decomp[p_idx + 1 :] + between_ops
+        else:
+            generator, multiplier = op.generator
+
+        generator = generator(wires)
+
+        diff_circuit = QuantumTape()
+        diff_circuit._ops = [copy(op).inv() for op in between_ops[::-1]] + [generator] + between_ops
+
+        # set the simulator state to be the pre-measurement state
+        device._state = self._state
+
+        # evolve the pre-measurement state under this new circuit
+        device.execute(diff_circuit)
+        dstate = device._pre_rotated_state  # TODO: this will only work for QubitDevices

```

Comment:
This is something that should work for all statevector simulators with the tape/queuing upgrades (since there is a separate measurement queue)

Reply:
Yep! Will have to wait for the device refactor though